### PR TITLE
Eliminate dependecy on openeo-udf package

### PR DESCRIPTION
**Pull request by soxofaan**
Addresses issue #159/EP-3578:

moves the pure UDF execution parts from [openeo-udf](https://github.com/Open-EO/openeo-udf) package (so not the server implementation aspects) to openeo Python client lib

(PR still under construction at the moment, but already available for review)

### CODE REVIEW

File: openeo/udf/run_code.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        'shapely': shapely,
+        'math': math,
+        # 'FeatureCollection': FeatureCollection,
+        # 'SpatialExtent': SpatialExtent,
+        'StructuredData': StructuredData,
+        # 'MachineLearnModel': MachineLearnModelConfig,
+        'DataCube': XarrayDataCube,
+        'XarrayDataCube': XarrayDataCube,
+        'UdfData': UdfData
+    }
```
Comment (soxofaan) [inline openeo/udf/run_code.py:36]: Improve autodetection of what is available?

**Comment by soxofaan** [comment]
Overview of what is added/changed:

- copy  and fine-tune `openeo_udf.api.datacube/DataCube` to `openeo.udf.xarraydatacube.XarrayDataCube`: I made name more explicit to avoid confusion with existing `DataCube`.
- copy and fine-tune `openeo_udf.api.run_code` to `openeo.udf.run_code` 
- copy and fine-tune `openeo_udf.api.structured_data` to `openeo.udf.structured_data`
- copy and fine-tune `openeo_udf.api.udf_data` to `openeo.udf.udf_data`  (not ported: feature_collection_list, ml_model_list, metadata, server_context)
- copy doctests/unittest and adapted to openeo python client lib style
- port/adapt and fix some unit tests that already existed in openeo python client lib

File: openeo/udf/run_code.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+                and _annotation_is_udf_datacube(params["cube"].annotation)
+                and _annotation_is_udf_datacube(sig.return_annotation)
+        ):
+            # found a datacube mapping function
+            if len(data.get_datacube_list()) != 1:
+                raise ValueError("The provided UDF expects exactly one datacube, but {c} were provided.".format(
+                    c=len(data.get_datacube_list())
+                ))
+            # TODO: also support calls without user context?
+            result_cube = func(data.get_datacube_list()[0], data.user_context)
```
Comment (soxofaan) [inline openeo/udf/run_code.py:175]: This code path is not fully covered with tests

**Comment by jdries** [comment]
I believe we also had this file:
openeo-udf/src/openeo_udf/api/udf_signatures.py
which was meant as documentation, to define allowed signatures supported by 'run_code'. Is it possible that this is not yet included, or is there some alternative?

**Comment by soxofaan** [comment]
>  openeo-udf/src/openeo_udf/api/udf_signatures.py

indeed, that is still to do (I though I did put a TODO note in `docs/udf.rst` about it, but apparently that didn't get committed). 

`udf_signatures` doesn't seem completely in sync with the actual implementation however, so that as to be tackled as well

**Comment by aljacob** [comment]
@clausmichele can you have a look if this might break our UDF deployment for python at the moment?

**Comment by clausmichele** [comment]
Hi Stefaan, currently the python UDF instance at Eurac is not running, but we will anyway have to update it in the future.
I think there will be no problems for us!

**Comment by soxofaan** [comment]
latest changes:
- port `openeo_udf.api.udf_signatures` to `openeo.udf.udf_signatures` 
- added `def apply_udf_data(data: UdfData)` that was supported but not described in `udf_signatures`
- bring back `_save_DataArray_to_JSON`/`_save_DataArray_to_NetCDF` which were used apparently in openeo-geopyspark-driver
- port `openeo_udf.api.feature_collection.FeatureCollection` to `openeo.udf.feature_collection.FeatureCollection` and simplified its API

File: docs/udf.rst [hunk 3]
```diff
@@ -13,25 +13,35 @@ a piece of data as possible.
...
 
 However, when you have a large piece of code that is hard to transform into predefined openEO functions,
 then it makes sense to use the UDF functionality.
 
 The section below gives an example to get you started.
 
+
+.. note::
+
+    Don not confuse user-defined profunctions (often abbreviated as UDF) with
```
Comment (soxofaan) [inline docs/udf.rst:33]: typo

File: docs/udf.rst [hunk 4]
```diff
@@ -40,59 +50,52 @@ To ensure that openEO understand your function, it needs to follow some rules, t
...
+    smoothing_udf = load_udf('smooth_savitzky_golay.py')
 
-    >>> ... # preparing the process
-    
-    >>> myprocess.download('test_input.nc', format='NetCDF')
-    
-    >>> udfstr=load_udf('../examples/udf/smooth_savitzky_golay.py')
-    >>> DataCube.execute_local_udf(udfstr, 'test_input.nc', fmt='netcdf'):
+    from openeo.udf import execute_local_udf
+    execute_local_udf(smoothing_udf, 'test_input.nc', fmt='netcdf'):
```
Comment (soxofaan) [inline docs/udf.rst:98]: trailing colon ?

File: examples/ndvi_composite.py [hunk 5]
```diff
@@ -104,7 +104,7 @@ def test_debug_udf_direct_invoke():
     https://open-eo.github.io/openeo-python-client/udf.html#example-downloading-a-datacube-and-executing-an-udf-locally
     depends on composite.nc file created in earlier function!
     """
-    from openeo.rest.conversions import datacube_from_file
-    udf_cube = datacube_from_file('masked.nc', fmt='netcdf')
+    from openeo.udf import XarrayDataCube
+    udf_cube = XarrayDataCube.from_file('masked.nc', fmt='netcdf')
```
Comment (soxofaan) [inline examples/ndvi_composite.py:108]: autodetect netcdf from `.nc` extension?

File: examples/udf/median_composite.py [hunk 6]
```diff
@@ -1,19 +1,16 @@
...
+def apply_datacube(cube: XarrayDataCube, context: dict) -> XarrayDataCube:
     """
     Applies a rolling window median composite to a timeseries datacube.
     This UDF preserves dimensionality, and assumes a datacube with a temporal dimension 't' as input.
     """
 
     array: xarray.DataArray = cube.get_array()
 
-    import pandas as pd
     import numpy as np
```
Comment (soxofaan) [inline examples/udf/median_composite.py:14]: move numpy import up

File: openeo/rest/conversions.py [hunk 7]
```diff
@@ -1,13 +1,10 @@
...
 Helpers for data conversions between Python ecosystem data types and openEO data structures.
 """
-import collections
-import json
-from typing import Union
 
 import numpy as np
 import pandas
-import xarray
+from deprecated import deprecated
```
Comment (soxofaan) [inline openeo/rest/conversions.py:7]: Add this module to sphinx documentation (and use sphinx `deprecated` decorator)

File: openeo/rest/conversions.py [hunk 8]
```diff
@@ -87,235 +84,29 @@ def timeseries_json_to_pandas(timeseries: dict, index: str = "date", auto_collap
...
-                ax.set_xticklabels([])
-                ax.set_yticklabels([])
-            if show_bandnames:
-                if i==0: ax.text(0.5,1.08, data.bands.values[j]+" ("+str(data.dtype)+")", size=fs, va="center", ha="center", transform=ax.transAxes)
-            if show_dates:
-                if j==0: ax.text(-0.08,0.5, data.t.dt.strftime("%Y-%m-%d").values[i], size=fs, va="center", ha="center", rotation=90,  transform=ax.transAxes)
 
-    if title is not None:
-        fig.text(0.,1.,title.split('/')[-1], size=fs, va="top", ha="left",weight='bold')
+@deprecated("Use :py:method:`XarrayIO.to_json_file` instead", version="0.6.2")
```
Comment (soxofaan) [inline openeo/rest/conversions.py:103]: move these up to original location

File: openeo/rest/datacube.py [hunk 9]
```diff
@@ -1446,51 +1445,10 @@ def execute(self) -> Dict:
...
-        # signature: UdfData(proj, datacube_list, feature_collection_list, structured_data_list, ml_model_list, metadata)
-        
-        # run the udf through the same routine as it would have been parsed in the backend
-        if udf_data is not None:
-            result=run_user_code(udf, udf_data)
-            return result
-        
-        return None
+    @deprecated(reason="Use :py:func:`openeo.udf.run_code.execute_local_udf` instead", version="0.6.2")
+    def execute_local_udf(udf: str, datacube: Union[str, 'xarray.DataArray', 'XarrayDataCube'] = None, fmt='netcdf'):
```
Comment (soxofaan) [inline openeo/rest/datacube.py:1449]: add minimal doc block for API docs

File: openeo/udf/feature_collection.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
+from typing import Optional, Union, Any, List
```
Comment (soxofaan) [inline openeo/udf/feature_collection.py:1]: Add note about origin of this implementation (openeo-udf project)

File: openeo/udf/run_code.py [hunk 11]
```diff
@@ -0,0 +1,232 @@
+import functools
```
Comment (soxofaan) [inline openeo/udf/run_code.py:1]: Add note about origin of this implementation (openeo-udf project)

File: openeo/udf/run_code.py [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+
+
+def _annotation_is_udf_data(annotation) -> bool:
+    return annotation is UdfData or _get_annotation_str(annotation) in {
+        _get_annotation_str(UdfData),
+        'openeo_udf.api.udf_data.UdfData'  # Legacy `openeo_udf` annotation
+    }
+
+
+def apply_timeseries_example(series: Series, context: Dict) -> Series:
```
Comment (soxofaan) [inline openeo/udf/run_code.py:90]: Is it necessary to have this example if we already have udf_signatures?

File: openeo/udf/run_code.py [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+    :param series:
+    :param context:
+    :return:
+    """
+    return series
+
+
+def _apply_timeseries_xarray(
+        array: xarray.DataArray,
+        callback: Callable[[Series], Series] = apply_timeseries_example,
```
Comment (soxofaan) [inline openeo/udf/run_code.py:103]: why a default value for `callback`?

File: openeo/udf/run_code.py [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+    applied = applied.reshape(orig_shape)
+    applied = numpy.moveaxis(applied, -1, time_position)
+    assert applied.shape == array.shape
+
+    return xarray.DataArray(applied, coords=array.coords, dims=array.dims, name=array.name)
+
+
+def apply_timeseries_generic(
+        udf_data: UdfData,
+        callback: Callable[[Series, dict], Series] = apply_timeseries_example
```
Comment (soxofaan) [inline openeo/udf/run_code.py:132]: why default for `callback`?

File: openeo/udf/run_code.py [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+    ]
+    # Insert the new tiles as list of raster collection tiles in the input object. The new tiles will
+    # replace the original input tiles.
+    udf_data.set_datacube_list(datacubes)
+    return udf_data
+
+
+def run_udf_code(code: str, data: UdfData) -> UdfData:
+    # TODO: current implementation uses first match directly, first check for multiple matches?
+    # TODO: more test coverage for this
```
Comment (soxofaan) [inline openeo/udf/run_code.py:154]: remove TODO about more test coverage

File: openeo/udf/run_code.py [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+        first_param = next(iter(params.values()), None)
+
+        if (
+                fn_name == 'apply_timeseries'
+                and 'series' in params and 'context' in params
+                and _annotation_is_pandas_series(params["series"].annotation)
+                and _annotation_is_pandas_series(sig.return_annotation)
+        ):
+            # this is a UDF that transforms pandas series
+            return apply_timeseries_generic(data, func)
```
Comment (soxofaan) [inline openeo/udf/run_code.py:173]: add logging of which UDF type was found

File: openeo/udf/structured_data.py [hunk 17]
```diff
@@ -0,0 +1,36 @@
+import builtins
```
Comment (soxofaan) [inline openeo/udf/structured_data.py:1]: Add note about origin of this implementation (openeo-udf project)

File: openeo/udf/udf_data.py [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+        """
+        The constructor of the UDF argument class that stores all data required by the
+        user defined function.
+
+        :param proj: A dictionary of form {"proj type string": "projection description"} i. e. {"EPSG":4326}
+        :param datacube_list: A list of data cube objects
+        :param feature_collection_list: A list of VectorTile objects
+        :param structured_data_list: A list of structured data objects
+        """
+        # TODO: when None is given: assign empty list or just None to keep things simple?
```
Comment (soxofaan) [inline openeo/udf/udf_data.py:33]: remove all the boilerplate related to this and simplify the implementation

File: openeo/udf/udf_data.py [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+
+        :param structured_data_list: A list of StructuredData objects
+        """
+        self._structured_data_list = structured_data_list or []
+
+    def del_structured_data_list(self):
+        """Delete all structured data entries"""
+        self._structured_data_list = None
+
+    # TODO: get rid of these getter/setter properties to make things more explicit and readable?
```
Comment (soxofaan) [inline openeo/udf/udf_data.py:95]: API with explicit getters and setters will be cleaner and more self-documenting: so maybe better to remove `property` stuff for now

File: openeo/udf/xarraydatacube.py [hunk 20]
```diff
@@ -0,0 +1,353 @@
+import collections
```
Comment (soxofaan) [inline openeo/udf/xarraydatacube.py:1]: add note about origin of this module

File: examples/ndvi_composite.py (hunk 5 above)
Comment (soxofaan) [reply examples/ndvi_composite.py:108]: done

File: openeo/udf/udf_data.py (hunk 19 above)
Comment (soxofaan) [reply openeo/udf/udf_data.py]: will keep getters and setters, but remove "deller"

**Pull request merged**